// marko/dist/runtime/vdom/AsyncVDOMBuilder.js
var EventEmitter = require("events-light");
var vdom = require("./vdom");
var VElement = vdom.aS_;
var VDocumentFragment = vdom.aT_;
var VComment = vdom.aU_;
var VText = vdom.aV_;
var VComponent = vdom.aW_;
var VFragment = vdom.aX_;
var virtualizeHTML = vdom.aY_;
var RenderResult = require("../RenderResult");
var defaultDocument = vdom.aZ_;
var morphdom = require("./morphdom");
var attrsHelper = require("./helpers/attrs");

var EVENT_UPDATE = "update";
var EVENT_FINISH = "finish";

function State(tree) {
    this.b__ = new EventEmitter();
    this.ba_ = tree;
    this.bb_ = false;
}

function AsyncVDOMBuilder(globalData, parentNode, parentOut) {
    if (!parentNode) {
        parentNode = new VDocumentFragment();
    }

    var state;

    if (parentOut) {
        state = parentOut.K_;
    } else {
        state = new State(parentNode);
    }

    this.bc_ = 1;
    this.bd_ = 0;
    this.be_ = null;
    this.bf_ = parentOut;

    this.data = {};
    this.K_ = state;
    this.az_ = parentNode;
    this.global = globalData || {};
    this.bg_ = [parentNode];
    this.bh_ = false;
    this.bi_ = undefined;
    this.i_ = null;

    this.l_ = null;
    this.ao_ = null;
    this.ax_ = null;
}

var proto = AsyncVDOMBuilder.prototype = {
    aO_: true,
    Y_: defaultDocument,

    bc: function (component, key, ownerComponent) {
        var vComponent = new VComponent(component, key, ownerComponent);
        return this.bj_(vComponent, 0, true);
    },

    q_: function (component, key, ownerComponent) {
        var vComponent = new VComponent(component, key, ownerComponent, true);
        this.bj_(vComponent, 0);
    },

    bj_: function (child, childCount, pushToStack) {
        this.az_.bk_(child);
        if (pushToStack === true) {
            this.bg_.push(child);
            this.az_ = child;
        }
        return childCount === 0 ? this : child;
    },

    element: function (tagName, attrs, key, component, childCount, flags, props) {
        var element = new VElement(tagName, attrs, key, component, childCount, flags, props);
        return this.bj_(element, childCount);
    },

    aL_: function (tagName, attrs, key, component, childCount, flags, props) {
        return this.element(tagName, attrsHelper(attrs), key, component, childCount, flags, props);
    },

    n: function (node, component) {
        // NOTE: We do a shallow clone since we assume the node is being reused
        //       and a node can only have one parent node.
        var clone = node.__();
        this.node(clone);
        clone.aB_ = component;

        return this;
    },

    node: function (node) {
        this.az_.bk_(node);
        return this;
    },

    text: function (text) {
        var type = typeof text;

        if (type != "string") {
            if (text == null) {
                return;
            } else if (type === "object") {
                if (text.toHTML) {
                    return this.h(text.toHTML());
                }
            }

            text = text.toString();
        }

        this.az_.bk_(new VText(text));
        return this;
    },

    comment: function (comment) {
        return this.node(new VComment(comment));
    },

    html: function (html) {
        if (html != null) {
            var vdomNode = virtualizeHTML(html, this.Y_ || document);
            this.node(vdomNode);
        }

        return this;
    },

    beginElement: function (tagName, attrs, key, component, childCount, flags, props) {
        var element = new VElement(tagName, attrs, key, component, childCount, flags, props);
        this.bj_(element, childCount, true);
        return this;
    },

    aJ_: function (tagName, attrs, key, component, childCount, flags, props) {
        return this.beginElement(tagName, attrsHelper(attrs), key, component, childCount, flags, props);
    },

    aM_: function (key, component, preserve) {
        var fragment = new VFragment(key, component, preserve);
        this.bj_(fragment, null, true);
        return this;
    },

    aN_: function () {
        this.endElement();
    },

    endElement: function () {
        var stack = this.bg_;
        stack.pop();
        this.az_ = stack[stack.length - 1];
    },

    end: function () {
        this.az_ = undefined;

        var remaining = --this.bc_;
        var parentOut = this.bf_;

        if (remaining === 0) {
            if (parentOut) {
                parentOut.bl_();
            } else {
                this.bm_();
            }
        } else if (remaining - this.bd_ === 0) {
            this.bn_();
        }

        return this;
    },

    bl_: function () {
        var remaining = --this.bc_;

        if (remaining === 0) {
            var parentOut = this.bf_;
            if (parentOut) {
                parentOut.bl_();
            } else {
                this.bm_();
            }
        } else if (remaining - this.bd_ === 0) {
            this.bn_();
        }
    },

    bm_: function () {
        var state = this.K_;
        state.bb_ = true;
        state.b__.emit(EVENT_FINISH, this.aP_());
    },

    bn_: function () {
        var lastArray = this._last;

        var i = 0;

        function next() {
            if (i === lastArray.length) {
                return;
            }
            var lastCallback = lastArray[i++];
            lastCallback(next);

            if (!lastCallback.length) {
                next();
            }
        }

        next();
    },

    error: function (e) {
        try {
            this.emit("error", e);
        } finally {
            // If there is no listener for the error event then it will
            // throw a new Error here. In order to ensure that the async fragment
            // is still properly ended we need to put the end() in a `finally`
            // block
            this.end();
        }

        return this;
    },

    beginAsync: function (options) {
        if (this.bh_) {
            throw Error("Tried to render async while in sync mode. Note: Client side await is not currently supported in re-renders (Issue: #942).");
        }

        var state = this.K_;

        if (options) {
            if (options.last) {
                this.bd_++;
            }
        }

        this.bc_++;

        var documentFragment = this.az_.bo_();
        var asyncOut = new AsyncVDOMBuilder(this.global, documentFragment, this);

        state.b__.emit("beginAsync", {
            out: asyncOut,
            parentOut: this
        });

        return asyncOut;
    },

    createOut: function () {
        return new AsyncVDOMBuilder(this.global);
    },

    flush: function () {
        var events = this.K_.b__;

        if (events.listenerCount(EVENT_UPDATE)) {
            events.emit(EVENT_UPDATE, new RenderResult(this));
        }
    },

    D_: function () {
        return this.K_.ba_;
    },

    aP_: function () {
        return this.bp_ || (this.bp_ = new RenderResult(this));
    },

    on: function (event, callback) {
        var state = this.K_;

        if (event === EVENT_FINISH && state.bb_) {
            callback(this.aP_());
        } else if (event === "last") {
            this.onLast(callback);
        } else {
            state.b__.on(event, callback);
        }

        return this;
    },

    once: function (event, callback) {
        var state = this.K_;

        if (event === EVENT_FINISH && state.bb_) {
            callback(this.aP_());
        } else if (event === "last") {
            this.onLast(callback);
        } else {
            state.b__.once(event, callback);
        }

        return this;
    },

    emit: function (type, arg) {
        var events = this.K_.b__;
        switch (arguments.length) {
            case 1:
                events.emit(type);
                break;
            case 2:
                events.emit(type, arg);
                break;
            default:
                events.emit.apply(events, arguments);
                break;
        }
        return this;
    },

    removeListener: function () {
        var events = this.K_.b__;
        events.removeListener.apply(events, arguments);
        return this;
    },

    sync: function () {
        this.bh_ = true;
    },

    isSync: function () {
        return this.bh_;
    },

    onLast: function (callback) {
        var lastArray = this._last;

        if (lastArray === undefined) {
            this._last = [callback];
        } else {
            lastArray.push(callback);
        }

        return this;
    },

    C_: function (doc) {
        var node = this.bi_;
        if (!node) {
            var vdomTree = this.D_();
            // Create the root document fragment node
            doc = doc || this.Y_ || document;
            this.bi_ = node = vdomTree.bq_(doc, null);
            morphdom(node, vdomTree, doc, this.i_);
        }
        return node;
    },

    toString: function (doc) {
        var docFragment = this.C_(doc);
        var html = "";

        var child = docFragment.firstChild;
        while (child) {
            var nextSibling = child.nextSibling;
            if (child.nodeType != 1) {
                var container = docFragment.ownerDocument.createElement("div");
                container.appendChild(child.cloneNode());
                html += container.innerHTML;
            } else {
                html += child.outerHTML;
            }

            child = nextSibling;
        }

        return html;
    },

    then: function (fn, fnErr) {
        var out = this;
        var promise = new Promise(function (resolve, reject) {
            out.on("error", reject).on(EVENT_FINISH, function (result) {
                resolve(result);
            });
        });

        return Promise.resolve(promise).then(fn, fnErr);
    },

    catch: function (fnErr) {
        return this.then(undefined, fnErr);
    },

    isVDOM: true,

    c: function (componentDef, key, customEvents) {
        this.l_ = componentDef;
        this.ao_ = key;
        this.ax_ = customEvents;
    }
};

proto.e = proto.element;
proto.be = proto.beginElement;
proto.ee = proto.aK_ = proto.endElement;
proto.t = proto.text;
proto.h = proto.w = proto.write = proto.html;

module.exports = AsyncVDOMBuilder;
