// marko/dist/runtime/components/Component.js
"use strict";
/* jshint newcap:false */

var complain;

var domInsert = require("../dom-insert");
var defaultCreateOut = require("../createOut");
var getComponentsContext = require("./ComponentsContext").E_;
var componentsUtil = require("./util");
var componentLookup = componentsUtil.h_;
var emitLifecycleEvent = componentsUtil.F_;
var destroyNodeRecursive = componentsUtil.G_;
var EventEmitter = require("events-light");
var RenderResult = require("../RenderResult");
var SubscriptionTracker = require("listener-tracker");
var inherit = require("raptor-util/inherit");
var updateManager = require("./update-manager");
var morphdom = require("../vdom/morphdom");
var eventDelegation = require("./event-delegation");
var domData = require("./dom-data");
var componentsByDOMNode = domData.H_;
var CONTEXT_KEY = "__subtree_context__";

var slice = Array.prototype.slice;

var COMPONENT_SUBSCRIBE_TO_OPTIONS;
var NON_COMPONENT_SUBSCRIBE_TO_OPTIONS = {
    addDestroyListener: false
};

var emit = EventEmitter.prototype.emit;
var ELEMENT_NODE = 1;

function removeListener(removeEventListenerHandle) {
    removeEventListenerHandle();
}

function walkFragments(fragment) {
    var node;

    while (fragment) {
        node = fragment.firstChild;

        if (!node) {
            break;
        }

        fragment = node.fragment;
    }

    return node;
}

function handleCustomEventWithMethodListener(component, targetMethodName, args, extraArgs) {
    // Remove the "eventType" argument
    args.push(component);

    if (extraArgs) {
        args = extraArgs.concat(args);
    }

    var targetComponent = componentLookup[component.I_];
    var targetMethod = typeof targetMethodName === "function" ? targetMethodName : targetComponent[targetMethodName];
    if (!targetMethod) {
        throw Error("Method not found: " + targetMethodName);
    }

    targetMethod.apply(targetComponent, args);
}

function resolveKeyHelper(key, index) {
    return index ? key + "_" + index : key;
}

function resolveComponentIdHelper(component, key, index) {
    return component.id + "-" + resolveKeyHelper(key, index);
}

/**
 * This method is used to process "update_<stateName>" handler functions.
 * If all of the modified state properties have a user provided update handler
 * then a rerender will be bypassed and, instead, the DOM will be updated
 * looping over and invoking the custom update handlers.
 * @return {boolean} Returns true if if the DOM was updated. False, otherwise.
 */
function processUpdateHandlers(component, stateChanges, oldState) {
    var handlerMethod;
    var handlers;

    for (var propName in stateChanges) {
        if (stateChanges.hasOwnProperty(propName)) {
            var handlerMethodName = "update_" + propName;

            handlerMethod = component[handlerMethodName];
            if (handlerMethod) {
                (handlers || (handlers = [])).push([propName, handlerMethod]);
            } else {
                // This state change does not have a state handler so return false
                // to force a rerender
                return;
            }
        }
    }

    // If we got here then all of the changed state properties have
    // an update handler or there are no state properties that actually
    // changed.
    if (handlers) {
        // Otherwise, there are handlers for all of the changed properties
        // so apply the updates using those handlers

        handlers.forEach(function (handler) {
            var propertyName = handler[0];
            handlerMethod = handler[1];

            var newValue = stateChanges[propertyName];
            var oldValue = oldState[propertyName];
            handlerMethod.call(component, newValue, oldValue);
        });

        emitLifecycleEvent(component, "update");

        component.J_();
    }

    return true;
}

function checkInputChanged(existingComponent, oldInput, newInput) {
    if (oldInput != newInput) {
        if (oldInput == null || newInput == null) {
            return true;
        }

        var oldKeys = Object.keys(oldInput);
        var newKeys = Object.keys(newInput);
        var len = oldKeys.length;
        if (len !== newKeys.length) {
            return true;
        }

        for (var i = 0; i < len; i++) {
            var key = oldKeys[i];
            if (oldInput[key] !== newInput[key]) {
                return true;
            }
        }
    }

    return false;
}

var componentProto;

/**
 * Base component type.
 *
 * NOTE: Any methods that are prefixed with an underscore should be considered private!
 */
function Component(id) {
    EventEmitter.call(this);
    this.id = id;
    this.K_ = null;
    this.L_ = null;
    this.M_ = null;
    this.N_ = null;
    this.O_ = null; // Used to keep track of bubbling DOM events for components rendered on the server
    this.P_ = null;
    this.I_ = null;
    this.Q_ = null;
    this.R_ = undefined;
    this.S_ = false;
    this.T_ = undefined;

    this.U_ = false;
    this.V_ = false;
    this.W_ = false;
    this.X_ = false;

    this.Y_ = undefined;

    this.n_ = {};
    this.Z_ = undefined;
}

Component.prototype = componentProto = {
    ___: true,

    subscribeTo: function (target) {
        if (!target) {
            throw TypeError();
        }

        var subscriptions = this.M_ || (this.M_ = new SubscriptionTracker());

        var subscribeToOptions = target.___ ? COMPONENT_SUBSCRIBE_TO_OPTIONS : NON_COMPONENT_SUBSCRIBE_TO_OPTIONS;

        return subscriptions.subscribeTo(target, subscribeToOptions);
    },

    emit: function (eventType) {
        var customEvents = this.P_;
        var target;

        if (customEvents && (target = customEvents[eventType])) {
            var targetMethodName = target[0];
            var isOnce = target[1];
            var extraArgs = target[2];
            var args = slice.call(arguments, 1);

            handleCustomEventWithMethodListener(this, targetMethodName, args, extraArgs);

            if (isOnce) {
                delete customEvents[eventType];
            }
        }

        if (this.listenerCount(eventType)) {
            return emit.apply(this, arguments);
        }
    },
    getElId: function (key, index) {
        if (!key) {
            return this.id;
        }
        return resolveComponentIdHelper(this, key, index);
    },
    getEl: function (key, index) {
        if (key) {
            var resolvedKey = resolveKeyHelper(key, index);
            var keyedElement = this.n_["@" + resolvedKey];

            if (!keyedElement) {
                var keyedComponent = this.getComponent(resolvedKey);

                if (keyedComponent) {

                    return walkFragments(keyedComponent.L_);
                    // eslint-disable-next-line no-constant-condition
                }
            }

            return keyedElement;
        } else {
            return this.el;
        }
    },
    getEls: function (key) {
        key = key + "[]";

        var els = [];
        var i = 0;
        var el;
        while (el = this.getEl(key, i)) {
            els.push(el);
            i++;
        }
        return els;
    },
    getComponent: function (key, index) {
        var rootNode = this.n_[resolveKeyHelper(key, index)];
        if (/\[\]$/.test(key)) {
            rootNode = rootNode && rootNode[Object.keys(rootNode)[0]];
            // eslint-disable-next-line no-constant-condition
        }
        return rootNode && componentsByDOMNode.get(rootNode);
    },
    getComponents: function (key) {
        var lookup = this.n_[key + "[]"];
        return lookup ? Object.keys(lookup).map(function (key) {
            return componentsByDOMNode.get(lookup[key]);
        }).filter(Boolean) : [];
    },
    destroy: function () {
        if (this.U_) {
            return;
        }

        var root = this.L_;

        this._a_();

        var nodes = root.nodes;

        nodes.forEach(function (node) {
            destroyNodeRecursive(node);

            if (eventDelegation._b_(node) !== false) {
                node.parentNode.removeChild(node);
            }
        });

        root.detached = true;

        delete componentLookup[this.id];
        this.n_ = {};
    },

    _a_: function () {
        if (this.U_) {
            return;
        }

        emitLifecycleEvent(this, "destroy");
        this.U_ = true;

        componentsByDOMNode.set(this.L_, undefined);

        this.L_ = null;

        // Unsubscribe from all DOM events
        this._c_();

        var subscriptions = this.M_;
        if (subscriptions) {
            subscriptions.removeAllListeners();
            this.M_ = null;
        }
    },

    isDestroyed: function () {
        return this.U_;
    },
    get state() {
        return this.K_;
    },
    set state(newState) {
        var state = this.K_;
        if (!state && !newState) {
            return;
        }

        if (!state) {
            state = this.K_ = new this._d_(this);
        }

        state._e_(newState || {});

        if (state.W_) {
            this._f_();
        }

        if (!newState) {
            this.K_ = null;
        }
    },
    setState: function (name, value) {
        var state = this.K_;

        if (!state) {
            state = this.K_ = new this._d_(this);
        }
        if (typeof name == "object") {
            // Merge in the new state with the old state
            var newState = name;
            for (var k in newState) {
                if (newState.hasOwnProperty(k)) {
                    state._g_(k, newState[k], true /* ensure:true */);
                }
            }
        } else {
            state._g_(name, value, true /* ensure:true */);
        }
    },

    setStateDirty: function (name, value) {
        var state = this.K_;

        if (arguments.length == 1) {
            value = state[name];
        }

        state._g_(name, value, true /* ensure:true */
        , true /* forceDirty:true */
        );
    },

    replaceState: function (newState) {
        this.K_._e_(newState);
    },

    get input() {
        return this.R_;
    },
    set input(newInput) {
        if (this.X_) {
            this.R_ = newInput;
        } else {
            this._h_(newInput);
        }
    },

    _h_: function (newInput, onInput, out) {
        onInput = onInput || this.onInput;
        var updatedInput;

        var oldInput = this.R_;
        this.R_ = undefined;
        this._i_ = out && out[CONTEXT_KEY] || this._i_;

        if (onInput) {
            // We need to set a flag to preview `this.input = foo` inside
            // onInput causing infinite recursion
            this.X_ = true;
            updatedInput = onInput.call(this, newInput || {}, out);
            this.X_ = false;
        }

        newInput = this.Q_ = updatedInput || newInput;

        if (this.W_ = checkInputChanged(this, oldInput, newInput)) {
            this._f_();
        }

        if (this.R_ === undefined) {
            this.R_ = newInput;
            if (newInput && newInput.$global) {
                this.T_ = newInput.$global;
            }
        }

        return newInput;
    },

    forceUpdate: function () {
        this.W_ = true;
        this._f_();
    },

    _f_: function () {
        if (!this.V_) {
            this.V_ = true;
            updateManager._j_(this);
        }
    },

    update: function () {
        if (this.U_ === true || this._k_ === false) {
            return;
        }

        var input = this.R_;
        var state = this.K_;

        if (this.W_ === false && state !== null && state.W_ === true) {
            if (processUpdateHandlers(this, state._l_, state._m_, state)) {
                state.W_ = false;
            }
        }

        if (this._k_ === true) {
            // The UI component is still dirty after process state handlers
            // then we should rerender

            if (this.shouldUpdate(input, state) !== false) {
                this._n_();
            }
        }

        this.J_();
    },

    get _k_() {
        return this.W_ === true || this.K_ !== null && this.K_.W_ === true;
    },

    J_: function () {
        this.W_ = false;
        this.V_ = false;
        this.Q_ = null;
        var state = this.K_;
        if (state) {
            state.J_();
        }
    },

    shouldUpdate: function () {
        return true;
    },

    F_: function (eventType, eventArg1, eventArg2) {
        emitLifecycleEvent(this, eventType, eventArg1, eventArg2);
    },

    _n_: function () {
        var self = this;
        var renderer = self._o_;

        if (!renderer) {
            throw TypeError();
        }

        var input = this.Q_ || this.R_;

        updateManager._p_(function () {
            self._q_(input, false).afterInsert(self.Y_);
        });

        this.J_();
    },

    _q_: function (input, isHydrate) {
        var doc = this.Y_;
        var globalData = this.T_;
        var rootNode = this.L_;
        var renderer = this._o_;
        var createOut = renderer.createOut || defaultCreateOut;
        var out = createOut(globalData);
        out.sync();
        out.Y_ = this.Y_;
        out[CONTEXT_KEY] = this._i_;

        var componentsContext = getComponentsContext(out);
        var globalComponentsContext = componentsContext.m_;
        globalComponentsContext._r_ = this;
        globalComponentsContext._s_ = isHydrate;

        renderer(input, out);

        var result = new RenderResult(out);

        var targetNode = out.D_().a_;

        morphdom(rootNode, targetNode, doc, componentsContext);

        return result;
    },

    _t_: function () {
        var root = this.L_;
        root.remove();
        return root;
    },

    _c_: function () {
        var eventListenerHandles = this.N_;
        if (eventListenerHandles) {
            eventListenerHandles.forEach(removeListener);
            this.N_ = null;
        }
    },

    get _u_() {
        var state = this.K_;
        return state && state._v_;
    },

    _w_: function (customEvents, scope) {
        var finalCustomEvents = this.P_ = {};
        this.I_ = scope;

        customEvents.forEach(function (customEvent) {
            var eventType = customEvent[0];
            var targetMethodName = customEvent[1];
            var isOnce = customEvent[2];
            var extraArgs = customEvent[3];

            finalCustomEvents[eventType] = [targetMethodName, isOnce, extraArgs];
        });
    },

    get el() {
        return walkFragments(this.L_);
    },

    get els() {
        return (this.L_ ? this.L_.nodes : []).filter(function (el) {
            return el.nodeType === ELEMENT_NODE;
        });
        // eslint-disable-next-line no-constant-condition
    }
};

componentProto.elId = componentProto.getElId;
componentProto._x_ = componentProto.update;
componentProto._y_ = componentProto.destroy;

// Add all of the following DOM methods to Component.prototype:
// - appendTo(referenceEl)
// - replace(referenceEl)
// - replaceChildrenOf(referenceEl)
// - insertBefore(referenceEl)
// - insertAfter(referenceEl)
// - prependTo(referenceEl)
domInsert(componentProto, function getEl(component) {
    return component._t_();
}, function afterInsert(component) {
    return component;
});

inherit(Component, EventEmitter);

module.exports = Component;
