// marko/dist/runtime/RenderResult.js
var domInsert = require("./dom-insert");
var complain;

function getComponentDefs(result) {
    var componentDefs = result.i_;

    if (!componentDefs) {
        throw Error("No component");
    }
    return componentDefs;
}

function RenderResult(out) {
    this.out = this.A_ = out;
    this.i_ = undefined;
}

module.exports = RenderResult;

var proto = RenderResult.prototype = {
    getComponent: function () {
        return this.getComponents()[0];
    },
    getComponents: function (selector) {
        if (this.i_ === undefined) {
            throw Error("Not added to DOM");
        }

        var componentDefs = getComponentDefs(this);

        var components = [];

        componentDefs.forEach(function (componentDef) {
            var component = componentDef.k_;
            if (!selector || selector(component)) {
                components.push(component);
            }
        });

        return components;
    },

    afterInsert: function (doc) {
        var out = this.A_;
        var componentsContext = out.i_;
        if (componentsContext) {
            this.i_ = componentsContext.B_(doc);
        } else {
            this.i_ = null;
        }

        return this;
    },
    getNode: function (doc) {
        return this.A_.C_(doc);
    },
    getOutput: function () {
        return this.A_.D_();
    },
    toString: function () {
        return this.A_.toString();
    },
    document: typeof document != "undefined" && document
};

Object.defineProperty(proto, "html", {
    get: function () {
        return this.toString();
        // eslint-disable-next-line no-constant-condition
    }
});

Object.defineProperty(proto, "context", {
    get: function () {
        return this.A_;
        // eslint-disable-next-line no-constant-condition
    }
});

// Add all of the following DOM methods to Component.prototype:
// - appendTo(referenceEl)
// - replace(referenceEl)
// - replaceChildrenOf(referenceEl)
// - insertBefore(referenceEl)
// - insertAfter(referenceEl)
// - prependTo(referenceEl)
domInsert(proto, function getEl(renderResult, referenceEl) {
    return renderResult.getNode(referenceEl.ownerDocument);
}, function afterInsert(renderResult, referenceEl) {
    var isShadow = typeof ShadowRoot === "function" && referenceEl instanceof ShadowRoot;
    return renderResult.afterInsert(isShadow ? referenceEl : referenceEl.ownerDocument);
});
